src/core/server/http/router/validator.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SchemaTypeError, ValidationError, ObjectType, Type, TypeOf } from '@kbn/config-schema';
+
+/**
+ * Allowed returned format of the custom validate function
+ * @public
+ */
+export type ValidateFunctionReturn<T> =
+  | {
+      value: T;
+      error?: undefined;
+    }
+  | {
+      value?: undefined;
+      error: SchemaTypeError;
+    };
+
+/**
+ * Custom validate function (only if @kbn/config-schema is not a valid option in your plugin)
+ * @public
+ */
+export type ValidateFunction<T> = (data: any) => ValidateFunctionReturn<T>;
```

Comment:
I wonder if we can be a little more restrictive than `any` here? I think it's a pain as 

https://github.com/elastic/kibana/blob/066613e2a6b839ad893aa61fb3624187042b5bd3/src/core/server/http/router/request.ts#L114-L127

`req.params`, `req.query`, `req.payload` have different types, but if possible I think at least something like `data: reqParamsType | reqQueryType | reqBodyType` is a little better (or is it just worse?)

Mega bonus point if we can context-type it depending on if its a query, params or body validation method, but not sure if possible.

Reply:
I guess the best we can assume here is `req.params` and `req.query` to always be `Record<string, string>`. `req.payload` is a bit harder to assume any type before validating it (although we've assumed `{}` up until now and it looks like it didn't harm, especially after Hapi's parsing). I'll see if it can be done without overcomplicating (even more) the typings